src/xref2/ref_tools.ml

```diff
@@ -61,183 +85,409 @@ let module_type_lookup_to_signature_lookup :
-        | `ClassType (id, t) ->
-            return
-              ( `Identifier
-                  (id :> Odoc_model.Paths_types.Identifier.reference_type),
-                `CT t ) )
-    | `Dot (parent, name) ->
-        resolve_label_parent_reference env parent
-        >>= signature_lookup_result_of_label_parent
-        >>= fun (parent', cp, sg) ->
-        let sg = Tools.prefix_signature (cp, sg) in
-        Find.type_in_sig sg name >>= fun t ->
-        return (`Type (parent', TypeName.of_string name), t)
-    | `Class (parent, name) -> (
-        resolve_signature_reference env parent >>= fun (parent', _, sg) ->
-        Find.type_in_sig sg (ClassName.to_string name) >>= function
-        | `C _ as c -> return (`Class (parent', name), c)
-        | _ -> None )
-    | `ClassType (parent, name) -> (
-        resolve_signature_reference env parent >>= fun (parent', cp, sg) ->
-        let sg = Tools.prefix_signature (cp, sg) in
-        Find.type_in_sig sg (ClassTypeName.to_string name) >>= function
-        | `CT _ as c -> return (`ClassType (parent', name), c)
-        | _ -> None )
-    | `Type (parent, name) -> (
-        resolve_signature_reference env parent >>= fun (parent', cp, sg) ->
-        let sg = Tools.prefix_signature (cp, sg) in
-        Find.type_in_sig sg (TypeName.to_string name) >>= function
-        | `T _ as c -> return (`Type (parent', name), c)
+let type_lookup_to_class_signature_lookup :
+    Env.t -> type_lookup_result -> class_signature_lookup_result option =
+  let resolved p' cs = Some ((p' :> Resolved.ClassSignature.t), cs) in
+  fun env -> function
+    | `T _ -> None
+    | `C (p', c) -> Tools.class_signature_of_class env c >>= resolved p'
+    | `CT (p', ct) -> Tools.class_signature_of_class_type env ct >>= resolved p'
+
+module M = struct
+  (** Module *)
+
+  type t = module_lookup_result
+
+  let of_component env m base_path' base_ref' : t =
+    let base_path, base_ref =
+      if m.Component.Module.hidden then (`Hidden base_path', `Hidden base_ref')
+      else (base_path', base_ref')
+    in
+    let p, r =
+      match Tools.get_module_path_modifiers env true m with
+      | None -> (base_path, base_ref)
+      | Some (`SubstAliased cp) ->
+          let cp = Tools.reresolve_module env cp in
+          let p = Lang_of.(Path.resolved_module empty cp) in
+          (`SubstAlias (cp, base_path), `SubstAlias (p, base_ref))
+      | Some (`Aliased cp) ->
+          let cp = Tools.reresolve_module env cp in
+          let p = Lang_of.(Path.resolved_module empty cp) in
+          (`Alias (cp, base_path), `SubstAlias (p, base_ref))
+      | Some (`SubstMT cp) ->
+          let cp = Tools.reresolve_module_type env cp in
+          (`Subst (cp, base_path), base_ref)
+    in
+    (r, p, m)
+
+  let in_signature env ((parent, parent_cp, sg) : signature_lookup_result) name
+      : t option =
+    let parent_cp = Tools.reresolve_parent env parent_cp in
+    let sg = Tools.prefix_signature (parent_cp, sg) in
```

Comment:
It is not necessary to prefix the whole signature, we could prefix the result of the lookup instead. I'll open a PR.

Reply:
Yes, we should do this everywhere, we're a bit inconsistent atm.